iceoryx_utils/test/moduletests/test_cxx_stack.cpp

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) 2021 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "iceoryx_utils/cxx/stack.hpp"
+#include "test.hpp"
+using namespace iox;
+using namespace ::testing;
+
+namespace
+{
+class TestClass
+{
+  public:
+    TestClass() noexcept = default;
+    TestClass(int a, int b, int c) noexcept
+        : m_a(a)
+        , m_b(b)
+        , m_c(c)
+    {
+    }
+
+    bool operator==(const TestClass& rhs) const noexcept
+    {
+        return m_a == rhs.m_a && m_b == rhs.m_b && m_c == rhs.m_c;
+    }
+    int m_a = 0, m_b = 0, m_c = 0;
+};
+
+class stack_test : public Test
+{
+  public:
+    static constexpr uint64_t STACK_SIZE = 10U;
+    cxx::stack<TestClass, STACK_SIZE> m_sut;
+};
+} // namespace
+
+TEST_F(stack_test, isEmptyOnCreation)
+{
+    EXPECT_THAT(m_sut.size(), Eq(0U));
+    EXPECT_THAT(m_sut.capacity(), Eq(STACK_SIZE));
+    EXPECT_THAT(m_sut.pop(), Eq(cxx::nullopt));
+}
+
+TEST_F(stack_test, pushingOneElementWithDefaultCtorSucceeds)
+{
+    ASSERT_TRUE(m_sut.push());
+    EXPECT_THAT(m_sut.size(), Eq(1U));
+    EXPECT_THAT(m_sut.capacity(), Eq(STACK_SIZE));
+
+    auto element = m_sut.pop();
+    ASSERT_TRUE(element.has_value());
+    EXPECT_THAT(*element, Eq(TestClass(0, 0, 0)));
+}
+
+TEST_F(stack_test, pushingOneElementWithCustomCtorSucceeds)
+{
+    ASSERT_TRUE(m_sut.push(1, 2, 3));
+    EXPECT_THAT(m_sut.size(), Eq(1U));
+    EXPECT_THAT(m_sut.capacity(), Eq(STACK_SIZE));
+
+    auto element = m_sut.pop();
+    ASSERT_TRUE(element.has_value());
+    EXPECT_THAT(*element, Eq(TestClass(1, 2, 3)));
+}
+
+TEST_F(stack_test, pushingElementsTillStackIsFullAndPoppingInLIFOOrderSucceeds)
```

Comment:
shouldn't this be two tests?

Reply:
I would say no! How do I verify that push has succeeded when I do not acquire the elements afterwards and inspect their values?!

I could move the first for loop in a separate test but then I would the same for loop again in the next test where I then acquire and check the elements.
Here I pursued also the technique to always check all side effects in every tests (the idea from the article you shared a month ago)